renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -0,0 +1,225 @@
+
+  for(size_t i = 0; i < modEvents.size(); i++)
+  {
+    driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+    driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_Full);
+
+    // read the post mod pixel value into the history event
+    PixelModification mod;
+    ModificationValue modValue;
+    PixelValue pixelValue;
+    driver->glReadPixels(x, y, 1, 1, eGL_RGBA, eGL_FLOAT, (void *)pixelValue.floatValue.data());
+    driver->glReadPixels(x, y, 1, 1, eGL_DEPTH_COMPONENT, eGL_FLOAT, (void *)&modValue.depth);
+    driver->glReadPixels(x, y, 1, 1, eGL_STENCIL_INDEX, eGL_INT, (void *)&modValue.stencil);
+    modValue.col = pixelValue;
+
+    RDCEraseEl(mod);
+    mod.eventId = modEvents[i].eventId;
+    mod.postMod = modValue;
+    history.push_back(mod);
+
+    if(i < modEvents.size() - 1)
+    {
+      driver->ReplayLog(modEvents[i].eventId + 1, modEvents[i + 1].eventId, eReplay_WithoutDraw);
+    }
+  }
+}
+};
+
+rdcarray<PixelModification> GLReplay::PixelHistory(rdcarray<EventUsage> events, ResourceId target,
+                                                   uint32_t x, uint32_t y, const Subresource &sub,
+                                                   CompType typeCast)
+{
+  rdcarray<PixelModification> history;
+
+  if(events.empty())
+    return history;
+
+  TextureDescription textureDesc = GetTexture(target);
+  if(textureDesc.format.type == ResourceFormatType::Undefined)
+    return history;
+
+  // When RenderDoc passes in the y, the value being passed in is with the Y axis starting from the
+  // top
+  // However, we need to have it starting from the bottom, so flip it by subtracting y from the
+  // height.
+  uint32_t flippedY = textureDesc.height - y - 1;
+
+  rdcstr regionName = StringFormat::Fmt(
+      "PixelHistory: pixel: (%u, %u) on %s subresource (%u, %u, %u) cast to %s with %zu events", x,
+      flippedY, ToStr(target).c_str(), sub.mip, sub.slice, sub.sample, ToStr(typeCast).c_str(),
+      events.size());
+
+  RDCDEBUG("%s", regionName.c_str());
+
+  uint32_t sampleIdx = sub.sample;
+
+  if(sampleIdx > textureDesc.msSamp)
+    sampleIdx = 0;
+
+  uint32_t sampleMask = ~0U;
+  if(sampleIdx < 32)
+    sampleMask = 1U << sampleIdx;
+
+  bool multisampled = (textureDesc.msSamp > 1);
+
+  if(sampleIdx == ~0U || !multisampled)
+    sampleIdx = 0;
+
+  GLPixelHistoryResources resources;
+
+  MakeCurrentReplayContext(&m_ReplayCtx);
+
+  PixelHistorySetupResources(m_pDriver, resources, textureDesc, sub, (uint32_t)events.size());
+
+  rdcarray<EventUsage> modEvents = QueryModifyingEvents(m_pDriver, resources, events, x, flippedY);
+
+  if(modEvents.empty())
+  {
+    return history;
+  }
+
+  QueryPostModPixelValues(m_pDriver, resources, modEvents, x, flippedY, history);
+
+  // copy the postMod to next history's preMod
+  for(size_t i = 1; i < history.size(); ++i)
+  {
+    history[i].preMod = history[i - 1].postMod;
+  }
+
+  PixelHistoryDestroyResources(m_pDriver, resources);
```

Comment:
This destroy will also need to happen above in the early `return history` if no modifying events are found.